### PR TITLE
Removed PlayerPaletteFromCurrentTileset

### PR DESCRIPTION
**Pull request by Mailaender**
<s>Use PlayerPaletteFromCurrentTileset for Tiberian Sun rather then two palettes with tileset filters.</s>

### CODE REVIEW

**Comment by reaperrr** [comment]
I haven't had time to look more closely, but according to Travis/lint this breaks a lot of palette references.

**Comment by Mailaender** [comment]
This is strange and unexpected as it works in-game.

**Comment by pchote** [comment]
It shouldn't work in game because this is not a player palette, but a normal palette called "player".  IIRC it is used for things that should be drawn in unittem/unitsno but not be remapped.

**Comment by Mailaender** [comment]
It is used as a base for the `PlayerColorPalette` which is remapped per player. I am not sure if I stumbled upon a bug here. The `PaletteReference` tags seem alright.

**Comment by reaperrr** [comment]
`PlayerPaletteFromCurrentTileset` sets the `PaletteDefinition` attribute to `IsPlayerPalette = true`, which sort of makes sense, given its name. This is what causes the lint errors.

In theory, since it's only a normal palette at this point and `PlayerColorPalette` does the same, I guess you could remove the `(true)` from `PlayerPaletteFromCurrentTileset`'s attribute, since the latter palette alone doesn't really do anything player-palette-specific apart from its name, so I'm not sure it should have that attribute set to true to begin with. You'd have to make it implement `IProvidesAssetBrowserPalettes` (like PaletteFromFile) to keep the palette(s) available in the asset browser then.

**Comment by pchote** [comment]
The trait should be renamed if it isn't really creating player palettes.

**Comment by reaperrr** [comment]
> The trait should be renamed if it isn't really creating player palettes.

We should rename the `PlayerPalette` field in the tileset definition as well then. Question is, to what. 
`ActorPalette`?

**Comment by pchote** [comment]
Hardcoding trait-specific data in the tilesets is awful to begin with.  Thinking about it, the cleanest solution would be to remove `PlayerPaletteFromCurrentTileset` completely and then use `PaletteFromFile` with tileset filters everywhere.

**Comment by Mailaender** [comment]
Went with a trait removal then.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 1]
```diff
@@ -953,6 +953,16 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
 					}
 				}
 
+				if (engineVersion < 20170831)
+				{
+					if (node.Key.StartsWith("PlayerPaletteFromCurrentTileset"))
+					{
+						node.Value.Nodes.Add(new MiniYamlNode("Tileset", ""));
+						node.Key = "PaletteFromFile";
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Shouldn't this carry over the old suffix?

Comment (Mailaender) [reply OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Yes, by just renaming the node key it should exactly do that.

Note: I haven't tested this upgrade rule as this trait isn't used anymore.

Comment (Mailaender) [reply OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Wait you mean suffix.

File: OpenRA.Game/Map/TileSet.cs [hunk 2]
```diff
@@ -185,7 +185,6 @@ public class TileSet
 		public readonly string Id;
 		public readonly int SheetSize = 512;
 		public readonly string Palette;
```
Comment (pchote) [inline OpenRA.Game/Map/TileSet.cs]: The same applies here too.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs (hunk 1 above)
Comment (Mailaender) [reply OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: You could only have one, so it will most likely not have a suffix.

**Comment by reaperrr** [comment]
Needs rebase.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs (hunk 1 above)
Comment (GraionDilach) [reply OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Then give it one because this might end up conflicting.

Comment (Mailaender) [reply OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: I found the `RenameNodeKey` function which should take care of this now.

**Comment by Mailaender** [comment]
Rebased.

File: OpenRA.Game/Map/TileSet.cs (hunk 2 above)
Comment (Mailaender) [reply OpenRA.Game/Map/TileSet.cs]: https://github.com/Mailaender/OpenRA/tree/move-palette-from-tileset will be a lot more invasive. I probably won't finish it.

**Comment by abcdefg30** [comment]
Needs another rebase... :/

File: OpenRA.Game/Map/TileSet.cs (hunk 2 above)
Comment (pchote) [reply OpenRA.Game/Map/TileSet.cs]: Ok, lets focus on just `PlayerPalette` here then.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 3]
```diff
@@ -974,6 +974,17 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
 					}
 				}
 
+				if (engineVersion < 20170831)
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Don't forget to update this date too

**Comment by Mailaender** [comment]
Rebased and added `StringComparison.Ordinal` which my new MonoDevelop 7 suggested.

**Review by pchote — APPROVED** [review]

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/3dbffd1f2b7c9230cf775a8f9649d66153c22a61...97bc9e26f7a962429c0731f8517b5637ebf7ccf7

**Pull request merged**